### PR TITLE
As a user, I can filter search result by a result count and it's amount

### PR DESCRIPTION
**Pull request by hanam1ni**
## What happened

Support filter with a number of results count

## Proof Of Work

Keywords can be filtered with a number of results count
image

### CODE REVIEW

File: lib/google_crawler/keywords.ex [hunk 1]
```diff
@@ -95,4 +98,20 @@ defmodule GoogleCrawler.Keywords do
...
+
+    case result_count_operation do
+      ">" -> where(keywords, [k], fragment("? > ?", k.result_count, ^value))
+      "<" -> where(keywords, [k], fragment("? < ?", k.result_count, ^value))
+      "=" -> where(keywords, [k], fragment("? = ?", k.result_count, ^value))
+      _ -> keywords
+    end
+  end
+
+  defp filter_by_result_count(keywords, _), do: keywords
```
Comment (andyduong1920) [inline lib/google_crawler/keywords.ex:116]: Should we group both `filter_by_result_count` functions together @hanam1ni ?

File: test/google_crawler/keywords_test.exs [hunk 2]
```diff
@@ -98,6 +102,57 @@ defmodule GoogleCrawler.KeywordsTest do
...
 
+  describe "list_keywords/1 with result count filtered" do
+    test "filters keywords with search results more than the given amount if given > in operation" do
+      user = insert(:user)
+      keyword1 = insert(:keyword, user: user)
+      _keyword2 = insert(:keyword, user: user)
+      insert(:search_result, url: "example1.com", keyword: keyword1)
+      insert(:search_result, url: "example2.net", keyword: keyword1)
+
+      [keyword1_in_db] = Keywords.list_keywords(user.id, %{"result_count_operation" => ">", "result_count_value" => "1"})
```
Comment (andyduong1920) [inline test/google_crawler/keywords_test.exs:113]: It hards to know where does the `result_count_value` comes from, should we add it into the `insert(:search_result, url: "example1.com", keyword: keyword1)` so it becomes more clear for this test?

```
...
insert(:search_result, url: "example1.com", keyword: keyword1, result_count_value: 2)

[keyword1_in_db] = Keywords.list_keywords(user.id, %{"result_count_operation" => ">", "result_count_value" => "2"})

assert keyword1_in_db.id == keyword1.id
```

File: test/google_crawler_web/features/keywords/filter_keyword_test.exs [hunk 3]
```diff
@@ -46,4 +48,23 @@ defmodule GoogleCrawlerWeb.FilterKeywordTest do
...
+
+    session
+    |> login_as(user)
+    |> visit("/keyword")
+    |> click(css(@selectors[:button_keyword_filter]))
+    |> click(css(@selectors[:option_result_count_filter]))
+    |> fill_in(css(@selectors[:input_result_count_filter]), with: "2")
+    |> click(css(@selectors[:button_submit_filter]))
+    |> assert_has(css(@selectors[:keyword_table_cell], text: keyword1.title))
+    |> refute_has(css(@selectors[:keyword_table_cell], text: keyword2.title))
```
Comment (andyduong1920) [inline test/google_crawler_web/features/keywords/filter_keyword_test.exs:68]: Same with the comment above, it hard to know why we don't see the `keyword2.title`, better to specify that in 

```
insert(:search_result, keyword: keyword1)
insert(:search_result, keyword: keyword2)
``

File: test/google_crawler_web/features/keywords/filter_keyword_test.exs [hunk 4]
```diff
@@ -46,4 +48,23 @@ defmodule GoogleCrawlerWeb.FilterKeywordTest do
...
     |> refute_has(css(@selectors[:keyword_table_cell], text: keyword3.title))
   end
+
+  feature "filters keywords which search results match amount and operation by the given params", %{session: session} do
+    user = insert(:user)
+    keyword1 = insert(:keyword, user: user)
+    keyword2 = insert(:keyword, user: user)
+
+    insert(:search_result, keyword: keyword1)
+    insert(:search_result, keyword: keyword1)
```
Comment (andyduong1920) [inline test/google_crawler_web/features/keywords/filter_keyword_test.exs:58]: Double `search_result` for keyword1, was that intentional?

File: lib/google_crawler/keywords.ex (hunk 1 above)
Comment (hanam1ni) [reply lib/google_crawler/keywords.ex:116]: @andyduong1920 this function is needed, there is a case that params do not have `result_count_operation` or `result_count_value` so it not filter by this one 😁

File: test/google_crawler/keywords_test.exs (hunk 2 above)
Comment (hanam1ni) [reply test/google_crawler/keywords_test.exs:113]: `result_count_value` is not a column attribute of search result, it's a count of search results that belong to the keyword. 

For this test, it creates 2 search result count for `keyword1` and list the keyword with search result count more than 1 🙌

File: test/google_crawler_web/features/keywords/filter_keyword_test.exs (hunk 4 above)
Comment (hanam1ni) [reply test/google_crawler_web/features/keywords/filter_keyword_test.exs:58]: yes, to create a keyword with 2 search result and this test is filter keyword which contains 2 search result

File: test/google_crawler_web/features/keywords/filter_keyword_test.exs (hunk 3 above)
Comment (hanam1ni) [reply test/google_crawler_web/features/keywords/filter_keyword_test.exs:68]: same as I reply above, but for a clearer test, I will update the selector name to show that `equal` operator is selected

File: lib/google_crawler/keywords.ex (hunk 1 above)
Comment (andyduong1920) [reply lib/google_crawler/keywords.ex:116]: Oh before I saw **another function** in between these 2 functions

```
defp filter_by_result_count(keywords, %{"result_count_operation" => result_count_operation, ...}
```

and 

```
defp filter_by_result_count(keywords, _), do: keywords
```

so my comment was mean group both `defp filter_by_result_count` sit to next together, I wasn't mean `join` both functions together 😁

But so far, for now, there is no function in between these 2 functions anymore, maybe my 👓 was having some issue at that time 😀. All good now Micky, sorry for this comment 🙏

File: lib/google_crawler_web/templates/keyword/index.html.eex [hunk 5]
```diff
@@ -14,59 +14,78 @@
...
           <div class="form-group row">
-            <div class="col-4">
+            <div class="keyword-filter__item col-lg-4 col-6">
+              <label for="title">Keyword Title</label>
+              <input class="form-control" name="keyword_filter[title]" placeholder='"apple", "rocket"' value=<%= @params["keyword_filter"]["title"] %> >
+            </div>
+            <div class="keyword-filter__item col-lg-4 col-6">
               <label for="url">URL in result</label>
-              <input class="form-control" name="keyword_filter[url]" placeholder='e.g. "google", ".com"' value=<%= @params["url"] %> >
+              <input class="form-control" name="keyword_filter[url]" placeholder='e.g. "google", ".com"' value=<%= @params["keyword_filter"]["url"] %> >
```
Comment (andyduong1920) [inline lib/google_crawler_web/templates/keyword/index.html.eex:27]: It's small thing but we still need the `double quote` in `"google", ".com"` 🤔

File: lib/google_crawler_web/templates/keyword/index.html.eex [hunk 6]
```diff
@@ -14,59 +14,78 @@
...
   <div class="col-12">
-    <div class="collapse" id="keywordFilter">
+    <div class="collapse keyword-filter" id="keywordFilter">
       <div class="card card-body">
         <form>
           <div class="form-group row">
-            <div class="col-4">
+            <div class="keyword-filter__item col-lg-4 col-6">
+              <label for="title">Keyword Title</label>
+              <input class="form-control" name="keyword_filter[title]" placeholder='"apple", "rocket"' value=<%= @params["keyword_filter"]["title"] %> >
```
Comment (andyduong1920) [inline lib/google_crawler_web/templates/keyword/index.html.eex:23]: Same question with the double quotes in `"apple", "rocket"` 🤔

File: lib/google_crawler_web/templates/keyword/index.html.eex [hunk 7]
```diff
@@ -14,59 +14,78 @@
 </section>
 <section class="row mt-3">
   <div class="col-12">
-    <div class="collapse" id="keywordFilter">
+    <div class="collapse keyword-filter" id="keywordFilter">
```
Comment (andyduong1920) [inline lib/google_crawler_web/templates/keyword/index.html.eex:17]: How about moving this into a `_filter` partial @hanam1ni 🤔

File: test/google_crawler/keywords_test.exs (hunk 2 above)
Comment (andyduong1920) [reply test/google_crawler/keywords_test.exs:113]: Oh I see, so there will be `multiple search results` per keyword :)

It's ok but next time I would suggest putting the search result into the same keyword record or something like having 1 keyword record with only 1 result record, we could store the result as `array or json` that avoid creating too many DB records as for example like 1 keyword has 10 search results, with 100 keywords we could easily have 10_000 search result records.

File: test/google_crawler/keywords_test.exs [hunk 8]
```diff
@@ -86,6 +102,57 @@ defmodule GoogleCrawler.KeywordsTest do
...
+
+      assert keyword1_in_db.id == keyword1.id
+    end
+
+    test "filters keywords with search results less than the given amount if given < in operation" do
+      user = insert(:user)
+      keyword1 = insert(:keyword, user: user)
+      keyword2 = insert(:keyword, user: user)
+      insert(:search_result, url: "example1.com", keyword: keyword1)
+      insert(:search_result, url: "example2.net", keyword: keyword1)
```
Comment (andyduong1920) [inline test/google_crawler/keywords_test.exs:123]: should we create only 1 search result for this case as the test is `< 1` so creating only 1 search result could make the test more accurate

**Review by andyduong1920 — COMMENTED** [review]
Just a few small comments @hanam1ni 🙌

File: lib/google_crawler_web/templates/keyword/index.html.eex (hunk 7 above)
Comment (hanam1ni) [reply lib/google_crawler_web/templates/keyword/index.html.eex:17]: the partial refactoring is in this PR https://github.com/hanam1ni/google-crawler/pull/24 😁

File: test/google_crawler/keywords_test.exs (hunk 8 above)
Comment (hanam1ni) [reply test/google_crawler/keywords_test.exs:123]: Good idea 👍

**Review by andyduong1920 — APPROVED** [review]
LGTM @hanam1ni

**Review by bterone — APPROVED** [review]

File: lib/google_crawler/keywords.ex [hunk 9]
```diff
@@ -19,11 +19,14 @@ defmodule GoogleCrawler.Keywords do
   def list_keywords(user_id, filter_params) do
     compacted_params = compact_params(filter_params)
 
-    keyword_with_result_report()
+    subquery(
```
[the remark this answers is not in the document]
Comment (hanam1ni) [reply lib/google_crawler/keywords.ex:22]: @bterone the result count is a virtual column that will be populated in database query in  `keyword_with_result_report` function. 

So if we want to filter by virtual column we need to make it as subquery then we can use `where` to filter the column

**Pull request merged**